pmd-core/src/main/java/net/sourceforge/pmd/PMD.java

```diff
@@ -356,6 +356,9 @@ public static void processFiles(final PMDConfiguration configuration, final Rule
 
         sortFiles(configuration, files);
 
+        // Make sure the cache is listening for analysis results
+        ctx.getReport().addListener(configuration.getAnalysisCache());
```

Comment:
The listeners might be provided by external libraries/programs, when they call PMD - so they could potentially be not-thread-safe...
I think, we had this problem, once we switched on multi-threading by default....

The underlying problem is of course the API the PMD provides for integration: How to execute PMD and how to gather the results (violations including processing errors and suppressions). If we find a better way than these listeners, we can mark this as deprecated. Maven-pmd-plugin uses a custom renderer by now to collect the results and doesn't use the listener anymore....